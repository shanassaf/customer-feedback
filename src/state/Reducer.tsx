const Reducer = (state: State, action: Action): State => {
	switch (action.type) {
		case 'ADD_FEEDBACK':
			return {
				...state,
				feedback: [action.feedback, ...state.feedback],
			};
		default:
			return state;
	}
};

export default Reducer;
