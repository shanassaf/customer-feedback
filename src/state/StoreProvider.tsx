import React, { createContext, useReducer } from 'react';
import Reducer from './Reducer';

const initialState: State = { comments: [] };

const StoreProvider = ({ children }) => {
	const [state, dispatch] = useReducer(Reducer, initialState);
	return <Context.Provider value={{ state, dispatch }}>{children}</Context.Provider>;
};

export const Context = createContext<ContextType>(null);
export default StoreProvider;
