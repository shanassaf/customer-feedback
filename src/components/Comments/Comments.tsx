import React, { useContext } from 'react';
import styled from 'styled-components';
import { Context } from '../../state/StoreProvider';

const StyledTitle = styled.h2`
	font-size: 20px;
`;
const StyledSubTitle = styled.h2`
	font-size: 12px;
`;
const StyledComments = styled.ul`
	border-top: 1px solid black;
	border-bottom: 1px solid black;
	padding: 10px 20px;
	max-width: 1024px;
	margin: 0 auto;
	margin-top: 50px;
`;
const StyledComment = styled.li`
	list-style: none;
	font-size: 13px;
	border: 1px solid black;
	border-radius: 5px;
	padding: 10px 10px;
	margin: 20px 0;
	background: #c5dadd;
`;
const StyledItem = styled.div``;

export const Comments = (): JSX.Element => {
	const {
		state: { feedback },
	} = useContext(Context);

	return (
		<StyledComments>
			<StyledTitle>Comments section</StyledTitle>
			<StyledSubTitle>See the feedback that users have left here!</StyledSubTitle>
			{feedback.map((feedback) => {
				const { name, comment, rating, uuid } = feedback;
				return (
					<StyledComment key={uuid}>
						<StyledItem key={`${name}${uuid}`}>{`Name: ${name}`}</StyledItem>
						<StyledItem key={`${rating}${uuid}`}>{`Rating: ${rating}`}</StyledItem>
						<StyledItem key={`${comment}${uuid}`}>{`Comment: ${comment}`}</StyledItem>
					</StyledComment>
				);
			})}
		</StyledComments>
	);
};
