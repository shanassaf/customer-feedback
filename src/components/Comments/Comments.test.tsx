import React from 'react';
import { render } from '@testing-library/react';
import { Comments } from './Comments';
import { Context } from '../../state/StoreProvider';

describe('Comments', () => {
	const dispatch = jest.fn();
	const state = {
		feedback: [
			{
				name: 'name1',
				email: 'email1',
				comment: 'text1',
				rating: '1',
				uuid: 'uuid1',
			},
			{
				name: 'name2',
				email: 'email2',
				comment: 'text2',
				rating: '2',
				uuid: 'uuid2',
			},
		],
	} as State;
	test('component renders correctly', () => {
		const { asFragment } = render(
			<Context.Provider value={{ dispatch, state }}>
				<Comments />
			</Context.Provider>
		);
		expect(asFragment()).toMatchSnapshot();
	});
});
