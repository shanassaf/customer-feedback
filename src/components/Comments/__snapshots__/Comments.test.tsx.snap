// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments component renders correctly 1`] = `
<DocumentFragment>
  <ul
    class="sc-dlnjPT kFrWIN"
  >
    <h2
      class="sc-bdnylx dgWwPV"
    >
      Comments section
    </h2>
    <h2
      class="sc-gtssRu cIzEkl"
    >
      See the feedback that users have left here!
    </h2>
    <li
      class="sc-hKFyIo jvjKQM"
    >
      <div
        class="sc-eCApGN"
      >
        Name: name1
      </div>
      <div
        class="sc-eCApGN"
      >
        Rating: 1
      </div>
      <div
        class="sc-eCApGN"
      >
        Comment: text1
      </div>
    </li>
    <li
      class="sc-hKFyIo jvjKQM"
    >
      <div
        class="sc-eCApGN"
      >
        Name: name2
      </div>
      <div
        class="sc-eCApGN"
      >
        Rating: 2
      </div>
      <div
        class="sc-eCApGN"
      >
        Comment: text2
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
